REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [43]:
import pandas as pd

In [44]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [45]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [46]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [47]:
print(bus.describe(include = 'all'))

         business_id              name                 address           city  \
count    6315.000000              6315                    6315           6315   
unique           NaN              5692                    5534              1   
top              NaN  STARBUCKS COFFEE  24 WILLIE MAYS PLAZA    San Francisco   
freq             NaN                72                     123           6315   
mean    48562.416627               NaN                     NaN            NaN   
std     33825.711496               NaN                     NaN            NaN   
min        10.000000               NaN                     NaN            NaN   
25%      5891.500000               NaN                     NaN            NaN   
50%     65012.000000               NaN                     NaN            NaN   
75%     79197.000000               NaN                     NaN            NaN   
max     90239.000000               NaN                     NaN            NaN   

       state postal_code   

#### Question 2b: How many unique business IDs are there?  

In [48]:
bus['business_id'].describe()
#there are 6315 unique business IDs

count     6315.000000
mean     48562.416627
std      33825.711496
min         10.000000
25%       5891.500000
50%      65012.000000
75%      79197.000000
max      90239.000000
Name: business_id, dtype: float64

#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [49]:
bus['name'].value_counts()
#Top five common businesses by name: Starbucks, Peet's, McDonalds, SF Soup Company and Walgreens

STARBUCKS COFFEE                                  72
PEET'S COFFEE & TEA                               24
MCDONALDS                                         12
WALGREENS                                         11
SAN FRANCISCO SOUP COMPANY                        11
                                                  ..
MONTESACRO                                         1
NOIR LOUNGE                                        1
BASQUE HOTEL                                       1
VICTOR'S BAKERY                                    1
INTERCONTINENTAL SAN FRANCISCO 5TH FL. KITCHEN     1
Name: name, Length: 5692, dtype: int64

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [50]:
type('postal_code')

str

#### Question 3b

What are the unique values of postal_code?

In [51]:
bus['postal_code'].describe()

count      6056
unique       46
top       94110
freq        585
Name: postal_code, dtype: object

In [52]:
unique_zip = bus['postal_code'].unique()
print(unique_zip)

['94104' '94109' '94133' '94110' '94122' '94103' '94115' '94131' '94111'
 '94117' '94107' '94108' '94102' '94132' '94105' '94134' '94124' '94116'
 '94121' '94118' '94112' '94123' '94114' '94127' '941102019' '941' '94545'
 '94066' '941033148' nan '94158' '95105' '94013' '94130' 'CA' '92672'
 '94120' '94143' '94609' '94101' '00000' '94188' '94621' '94014' '94129'
 '94602' 'Ca']


In [53]:
type('postal_code')

str

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [54]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [55]:
validBiz = bus[bus['postal_code'].isin(validZip)]
validBiz.describe()
#5999 buisiness in the new df 'validZip'

business_id     latitude    longitude
count   5999.000000  3516.000000  3516.000000
mean   47485.367061    37.773820  -122.425356
std    33977.547969     0.022924     0.027374
min       10.000000    37.668824  -122.510896
25%     5798.000000    37.760526  -122.436389
50%    63943.000000    37.781118  -122.418515
75%    78780.000000    37.790124  -122.406193
max    90239.000000    37.808931  -122.376657

In [56]:
validZip

['94102',
 '94103',
 '94104',
 '94105',
 '94107',
 '94108',
 '94109',
 '94110',
 '94111',
 '94112',
 '94114',
 '94115',
 '94116',
 '94117',
 '94118',
 '94121',
 '94122',
 '94123',
 '94124',
 '94127',
 '94131',
 '94132',
 '94133',
 '94134']

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [57]:
null_long = validBiz[pd.isnull(validBiz['longitude'])]
null_long.describe()
#2,483 buisinesses are missing longitude values - they are also missing latitude

business_id  latitude  longitude
count   2483.000000       0.0        0.0
mean   76922.726943       NaN        NaN
std    16651.846611       NaN        NaN
min      821.000000       NaN        NaN
25%    75781.500000       NaN        NaN
50%    80627.000000       NaN        NaN
75%    85157.000000       NaN        NaN
max    90239.000000       NaN        NaN

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [58]:
not_null_long = validBiz[pd.notnull(validBiz['longitude'])]
not_null_long.describe()

business_id     latitude    longitude
count   3516.000000  3516.000000  3516.000000
mean   26696.696815    37.773820  -122.425356
std    27014.275398     0.022924     0.027374
min       10.000000    37.668824  -122.510896
25%     3321.500000    37.760526  -122.436389
50%     9278.000000    37.781118  -122.418515
75%    60933.750000    37.790124  -122.406193
max    71915.000000    37.808931  -122.376657

In [59]:
#Create two dataframes from validZip df above to merge into one df.
df1 = null_long[['business_id']].groupby(null_long['postal_code']).count().rename(columns={'business_id': 'null_lon'})
df2 = not_null_long[['business_id']].groupby(not_null_long['postal_code']).count().rename(columns={'business_id': 'not_null_lon'})
final_biz = pd.merge(df1, df2, on="postal_code", how="inner")
final_biz

null_lon  not_null_lon
postal_code                        
94102             196           262
94103             252           306
94104              59            74
94105             100           126
94107             247           204
94108              82           142
94109             146           237
94110             251           334
94111             115           167
94112              59           131
94114             108           122
94115              76           162
94116              35            58
94117              67           130
94118             100           143
94121              55           106
94122             123           147
94123              53           116
94124              88            76
94127              28            43
94131              14            34
94132              65            69
94133             135           282
94134              29            45

In [60]:
final_biz.describe()

null_lon  not_null_lon
count   24.000000      24.00000
mean   103.458333     146.50000
std     70.217661      85.03861
min     14.000000      34.00000
25%     58.000000      75.50000
50%     85.000000     130.50000
75%    126.000000     176.25000
max    252.000000     334.00000

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [61]:
#Other than using .describe to see that one zip has 252 businesses missing longitude, we have to perform basic
#math functions on the array. Divide null_long
final_biz['missing_long']= final_biz['null_lon'] / (final_biz['null_lon'] + final_biz['not_null_lon']) * 100

final_biz.sort_values('missing_long', ascending=False).head()

null_lon  not_null_lon  missing_long
postal_code                                      
94107             247           204     54.767184
94124              88            76     53.658537
94132              65            69     48.507463
94114             108           122     46.956522
94122             123           147     45.555556

# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [62]:
len(ins)

15430

In [63]:
ins.describe()

business_id         score          date
count  15430.000000  15430.000000  1.543000e+04
mean   40207.358004     91.179391  2.015071e+07
std    32795.456282      8.024274  8.720966e+03
min       10.000000     46.000000  2.013123e+07
25%     5215.000000     87.000000  2.014072e+07
50%    37773.500000     93.000000  2.015060e+07
75%    71057.000000     98.000000  2.016052e+07
max    89900.000000    100.000000  2.016122e+07

In [64]:
ins.nunique()
#5730 unique business ids

business_id    5730
score            49
date            814
type              2
dtype: int64

In [65]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

In [66]:
ins['score'].describe()
#15,430 entries in df

count    15430.000000
mean        91.179391
std          8.024274
min         46.000000
25%         87.000000
50%         93.000000
75%         98.000000
max        100.000000
Name: score, dtype: float64

#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

#### Exploring the data

In [67]:
#what are the unique 'types'
ins['type'].unique()

array(['routine', 'complaint'], dtype=object)

In [68]:
#how many "complaints" are there? 
#or this one: ins['type'].value_counts(ascending=True)
ins[['business_id','type']].groupby('type').count()

business_id
type                  
complaint            1
routine          15429

In [69]:
#new df with only entries without 'complaint'
ins2 = ins[(ins['type'] != 'complaint')]
ins2.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

In [70]:
#making sure that ins2 dropped complaint
ins2[['business_id','type']].groupby('type').count()

business_id
type                
routine        15429

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [71]:
#I tried a few methods and got mixed results. When I used 'ins2' it returned a giant red box 
#and said I was writing over a slice. I tried the datetime function. That gave me new rows, but more red boxes, 
#and suddenly all the dates were for 1970. I finally went back to the original csv and then it worked... which makes 
#me think that if I exported ins2 as csv, then read that in, maybe I'd have more luck? 

ins['year'] = ins['date'].astype(str).str[0:4]
ins.head()

business_id  score      date     type  year
0           10     82  20160503  routine  2016
1           10     94  20140729  routine  2014
2           10     92  20140114  routine  2014
3           19     94  20160513  routine  2016
4           19     94  20141110  routine  2014

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [72]:
ins['year'].unique()

array(['2016', '2014', '2015', '2013'], dtype=object)

In [73]:
ins['year'].value_counts()

2016    5840
2014    5629
2015    3923
2013      38
Name: year, dtype: int64

In [74]:
ins3 = ins[(ins['year'] != '2013')]
ins3.head()

business_id  score      date     type  year
0           10     82  20160503  routine  2016
1           10     94  20140729  routine  2014
2           10     92  20140114  routine  2014
3           19     94  20160513  routine  2016
4           19     94  20141110  routine  2014

In [75]:
ins3[['business_id','year']].groupby('year').count()

business_id
year             
2014         5629
2015         3923
2016         5840

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [76]:
ins3['year'].astype(str)
ins3[ins3['year'].str.contains('2016')]

business_id  score      date     type  year
0               10     82  20160503  routine  2016
3               19     94  20160513  routine  2016
6               24     98  20161005  routine  2016
7               24     96  20160311  routine  2016
10              31     84  20160816  routine  2016
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[5840 rows x 5 columns]

In [40]:
ins3[ins3['year'].str.contains('2016')].describe()

business_id        score          date
count   5840.000000  5840.000000  5.840000e+03
mean   44598.291438    90.233733  2.016068e+07
std    33999.580577     8.133708  2.998559e+02
min       10.000000    52.000000  2.016010e+07
25%     5619.750000    86.000000  2.016042e+07
50%    61064.500000    92.000000  2.016070e+07
75%    77445.750000    96.000000  2.016092e+07
max    89900.000000   100.000000  2.016122e+07

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [81]:
ins4 = ins3[ins3['year'].str.contains('2016')]

#pandas merge(bus df),(ins4 df), on the business_is (what they both have in common), on the left (which is bus_df)
scary_merge = pd.merge(bus, ins4, on ='business_id', how = 'outer')

scary_merge.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4           31      NORMAN'S ICE CREAM AND FREEZES   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
4           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   

    longitude  phone_number  score        date     type  year  
0 -122.403816  +14154217044   82.0  20160503.0  routine  2016  
1 -122.421547  +14157763262   94.0  20160513.0  routine  2016  
2 -122.403135  +14156779494   98.0  20161005.0  routine  2016  
3 -122.403135  +14156779494   96.0  20160311.0  routine  2016  
4 -122.419004           NaN   84.0  20160816.0  routine  2016

In [86]:
#When I got down to nsmallest, I saw that I had all naan values for score, which made me want to look at ins4. 
#I created ins4 wrong, so had to re-write. Just testing here. 
ins4.sample(10)

business_id  score      date     type  year
4363          5639     96  20160410  routine  2016
14125        81104     94  20160116  routine  2016
7974         39608     90  20160908  routine  2016
9591         65485     92  20160924  routine  2016
8697         61806     96  20161013  routine  2016
6711         26146    100  20161107  routine  2016
14577        82926     96  20161003  routine  2016
2080          2764     70  20160830  routine  2016
13564        79008    100  20160814  routine  2016
1406          1922     86  20161012  routine  2016

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [84]:
#df.nsmallest(3, 'population')
scary_merge.nsmallest(20,['score'])

business_id                         name              address  \
4892        71373      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD    
4873        71232                   UNCLE CAFE       65 WAVERLY PL    
6281        81667                 CRAZY PEPPER   2257 SAN JOSE AVE    
7325        87652                    POKI TIME     2101 LOMBARD ST    
3008        34585             HING WANG BAKERY        339 JUDAH ST    
5627        78328                   GOLDEN WOK   295 B ORIZABA AVE    
549          1843          MA'S DIM SUM & CAFE      1315 POWELL ST    
836          2716                 RED A BAKERY      634 CLEMENT ST    
4209        67330      SAN TUNG RESTAURANT LLC      1031 IRVING ST    
4903        71440  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST    
5383        76615                   TOP'S CAFE  2460 SAN BRUNO AVE    
6067        80591    L & G VIETNAMESE SANDWICH         602 EDDY ST    
6645        83647    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST    
624          2044          JOANIE'S DINER INC.   1329 COLUMBUS AVE    
3563        61161                  JADE GARDEN      3608 BALBOA ST    
4234        67564                        ITHAI         720 POST ST    
6938        85587         PAK NAM THAI CUISINE       655 LARKIN ST    
4850        70996             HONG KONG LOUNGE     5322 GEARY BLVD    
5013        72176                 WOLFES LUNCH        1220 16TH ST    
6151        81106          VALLARTA'S TACO BAR     2022 MISSION ST    

               city state postal_code   latitude   longitude  phone_number  \
4892  San Francisco    CA       94121  37.780144 -122.482040  +14156685105   
4873  San Francisco    CA       94108  37.794034 -122.406794  +14155701380   
6281  San Francisco    CA       94112        NaN         NaN  +14157348687   
7325  San Francisco    CA       94123        NaN         NaN  +14156366868   
3008  San Francisco    CA       94122  37.762174 -122.465809  +14156813928   
5627  San Francisco    CA       94132        NaN         NaN  +14155868000   
549   San Francisco    CA       94133  37.796683 -122.410134  +14157883532   
836   San Francisco    CA       94118  37.782888 -122.466101  +14155663888   
4209  San Francisco    CA       94122  37.763891 -122.468805  +14152420828   
4903  San Francisco    CA       94108        NaN         NaN  +14159567888   
5383  San Francisco    CA       94134        NaN         NaN  +14157156868   
6067  San Francisco    CA       94109        NaN         NaN  +14158856402   
6645  San Francisco    CA       94123        NaN         NaN  +14156665567   
624   San Francisco    CA       94133  37.806546 -122.419277  +14158232670   
3563  San Francisco    CA       94121  37.775728 -122.497612  +14152218168   
4234  San Francisco    CA       94109  37.787691 -122.413717  +14156581010   
6938  San Francisco    CA       94109        NaN         NaN  +14156097089   
4850  San Francisco    CA       94121  37.780538 -122.476632           NaN   
5013  San Francisco    CA       94107        NaN         NaN           NaN   
6151  San Francisco    CA       94110        NaN         NaN  +14152528810   

      score        date     type  year  
4892   52.0  20161031.0  routine  2016  
4873   55.0  20160308.0  routine  2016  
6281   55.0  20160901.0  routine  2016  
7325   55.0  20160721.0  routine  2016  
3008   56.0  20160201.0  routine  2016  
5627   56.0  20160824.0  routine  2016  
549    57.0  20160216.0  routine  2016  
836    57.0  20160323.0  routine  2016  
4209   57.0  20160825.0  routine  2016  
4903   57.0  20161121.0  routine  2016  
5383   57.0  20160504.0  routine  2016  
6067   57.0  20160129.0  routine  2016  
6645   58.0  20160714.0  routine  2016  
624    59.0  20160204.0  routine  2016  
3563   59.0  20160107.0  routine  2016  
4234   59.0  20161108.0  routine  2016  
6938   59.0  20160927.0  routine  2016  
4850   60.0  20160121.0  routine  2016  
5013   60.0  20160225.0  routine  2016  
6151   60.0  20160811.0  routine  2016

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!